### PR TITLE
Making /users as public and adding id as query param

### LINKED ISSUE #983: Fix the route in integration tests, for API `users/userId/:id` changing to `users/?id=abc`
As we are changing the API route the test cases are also to be updated as per the new route `users/?id=abc`

### PR DESCRIPTION
**Pull request by endeavourmonk**
fixes #983 
Dont' merge now before ui changes
- [X]  made /users api route public.
- [x]  added id in validators.
- [X] checking in users controller if it contains id query then returning the resulting user.
- [X] removed authentication from tests of /users and wrote test for query param id.

The other tasks are mentioned here https://github.com/Real-Dev-Squad/website-backend/issues/966

### CODE REVIEW

File: scripts/tests/testUnit.sh [hunk 1]
```diff
@@ -8,4 +8,4 @@ json=$(node -e "console.log(require('config').get('firestore'))")
 project_id=$(echo $json | grep -o '"project_id":[^,}]*' | cut -d':' -f2 | tr -d '"' | tr -d '[:space:]')
 
 echo 'Start firestore emulator and run unit tests:'
-firebase emulators:exec 'nyc --x=controllers --x=test --x=docs --x=mockdata mocha test/unit/**' --project=$project_id
+firebase emulators:exec 'nyc --x=controllers --x=test --x=docs --x=mockdata mocha test/unit/**' --project='rds-backend'
```
Comment (Muralidhar22) [inline scripts/tests/testUnit.sh:11]: I don't think this file meant to be changed

File: routes/users.js [hunk 2]
```diff
@@ -11,7 +11,8 @@ const { getUserBadges } = require("../controllers/badges");
 router.post("/verify", authenticate, users.verifyUser);
 router.get("/userId/:userId", users.getUserById);
 router.patch("/self", authenticate, userValidator.updateUser, users.updateSelf);
-router.get("/", authenticate, userValidator.getUsers, users.getUsers);
+// router.get("/", authenticate, userValidator.getUsers, users.getUsers);
+router.get("/", userValidator.getUsers, users.getUsers);
```
Comment (Muralidhar22) [inline routes/users.js:15]: why commented out and not removed? line-14

File: models/users.js [hunk 3]
```diff
@@ -136,6 +136,7 @@ const fetchPaginatedUsers = async (query) => {
     // INFO: https://github.com/Real-Dev-Squad/website-backend/pull/873#discussion_r1064229932
     const size = parseInt(query.size) || 100;
     const doc = (query.next || query.prev) && (await userModel.doc(query.next || query.prev).get());
+
```
Comment (Muralidhar22) [inline models/users.js:139]: remove this unnecessary new line.

File: routes/users.js [hunk 4]
```diff
@@ -11,7 +11,8 @@ const { getUserBadges } = require("../controllers/badges");
 router.post("/verify", authenticate, users.verifyUser);
 router.get("/userId/:userId", users.getUserById);
 router.patch("/self", authenticate, userValidator.updateUser, users.updateSelf);
-router.get("/", authenticate, userValidator.getUsers, users.getUsers);
+// router.get("/", authenticate, userValidator.getUsers, users.getUsers);
```
Comment (Muralidhar22) [inline routes/users.js:14]: why commented out and not removed?

**Comment by heyrandhir** [comment]
The tests are failing can you check @endeavourmonk.

File: scripts/tests/testIntegration.sh [hunk 5]
```diff
@@ -8,4 +8,4 @@ json=$(node -e "console.log(require('config').get('firestore'))")
 project_id=$(echo $json | grep -o '"project_id":[^,}]*' | cut -d':' -f2 | tr -d '"' | tr -d '[:space:]')
 
 echo 'Start firestore emulator and run integration tests:'
-firebase emulators:exec 'nyc mocha test/integration/**' --project=$project_id
+firebase emulators:exec 'nyc mocha test/integration/**' --project='rds-backend'
```
Comment (vikhyat187) [inline scripts/tests/testIntegration.sh:11]: @endeavourmonk need not hardcode this value

File: test/integration/users.test.js [hunk 6]
```diff
@@ -585,6 +575,39 @@ describe("Users", function () {
...
   });
 
+  describe("GET /users/?id", function () {
+    beforeEach(async function () {
+      await addJoinData(joinData(userId)[0]);
+    });
+    it("Should return given user by id", function (done) {
+      chai
+        .request(app)
+        .get(`/users/?id`)
```
Comment (vikhyat187) [inline test/integration/users.test.js:585]: I think the test cases are failing as no Id is being passed, @heyrandhir @RitikJaiswal75 correct me if wrong.

**Review by vikhyat187 — CHANGES_REQUESTED** [review]

**Comment by endeavourmonk** [comment]
Done all changes

File: test/integration/users.test.js [hunk 7]
```diff
@@ -585,6 +575,37 @@ describe("Users", function () {
...
+      chai
+        .request(app)
+        .get(`/users/?id=${userId}`)
+        .end((err, res) => {
+          if (err) {
+            return done(err);
+          }
+          expect(res).to.have.status(200);
+          expect(res.body).to.be.a("object");
+          expect(res.body.message).to.equal("User returned successfully!");
```
Comment (vikhyat187) [inline test/integration/users.test.js:589]: also add the check on phone number and email not being present, as in the older test cases of `users/userId/:id`

File: test/integration/users.test.js [hunk 8]
```diff
@@ -306,7 +297,7 @@ describe("Users", function () {
       chai
         .request(app)
         .get(`/users?page=1&size=2`)
-        .set("cookie", `${cookieName}=${jwt}`)
+
```
Comment (vikhyat187) [inline test/integration/users.test.js:300]: NIT: remove the extra line

File: test/integration/users.test.js (hunk 7 above)
Comment (vikhyat187) [reply test/integration/users.test.js:588]: This is not being added Ujjawal

File: routes/users.js [hunk 9]
```diff
@@ -11,7 +11,7 @@ const { getUserBadges } = require("../controllers/badges");
 router.post("/verify", authenticate, users.verifyUser);
 router.get("/userId/:userId", users.getUserById);
 router.patch("/self", authenticate, userValidator.updateUser, users.updateSelf);
-router.get("/", authenticate, userValidator.getUsers, users.getUsers);
+router.get("/", userValidator.getUsers, users.getUsers);
```
Comment (RitikJaiswal75) [inline routes/users.js:14]: Will we be able to send a body in get request? if yes how? if no why do we need the request data validator?

File: test/integration/users.test.js [hunk 10]
```diff
@@ -585,6 +574,38 @@ describe("Users", function () {
...
+        .request(app)
+        .get(`/users/?id=${userId}`)
+        .end((err, res) => {
+          if (err) {
+            return done(err);
+          }
+          expect(res).to.have.status(200);
+          expect(res.body).to.be.a("object");
+          expect(res.body.message).to.equal("User returned successfully!");
+          expect(res.body.user).to.be.a("object");
```
Comment (RitikJaiswal75) [inline test/integration/users.test.js:589]: shouldn't we try to get the data from db and check weather returned user id matches the query param id passed?

**Review by RitikJaiswal75 — CHANGES_REQUESTED** [review]

File: test/integration/users.test.js (hunk 7 above)
Comment (endeavourmonk) [reply test/integration/users.test.js:588]: @vikhyat187 updated test for checking that phone and email should not be present and deleted phone and email from user.

File: routes/users.js (hunk 9 above)
Comment (endeavourmonk) [reply routes/users.js:14]: HTTP GET method should not have a request body.

File: controllers/users.js [hunk 11]
```diff
@@ -70,6 +70,38 @@ const getUsers = async (req, res) => {
...
+      if (!result.userExists) {
+        return res.boom.notFound("User doesn't exist");
+      }
+
+      const { phone = "", email = "", ...user } = result.user;
+      try {
+        user.phone = obfuscate.obfuscatePhone(phone);
+        user.email = obfuscate.obfuscateMail(email);
+        delete user.phone;
+        delete user.email;
```
Comment (heyrandhir) [inline controllers/users.js:93]: why are you obfuscating if you have to delete 🤔

File: test/integration/users.test.js [hunk 12]
```diff
@@ -585,6 +574,40 @@ describe("Users", function () {
...
+          expect(res.body.user).to.not.have.property("phone");
+          expect(res.body.user).to.not.have.property("email");
```
Comment (heyrandhir) [inline test/integration/users.test.js:591]: you only checked the fields which were not present . what about the fields which will be present

**Review by heyrandhir — CHANGES_REQUESTED** [review]
can we also update the api contracts in accordance with this change.

File: test/integration/users.test.js (hunk 12 above)
Comment (endeavourmonk) [reply test/integration/users.test.js:591]: is it necessary to check all the fields, if yes then what fields need to be present.

File: test/integration/users.test.js (hunk 10 above)
Comment (endeavourmonk) [reply test/integration/users.test.js:589]: should we fetch all users and then check if user exist, with id passed in query param? Is this we should do.

Comment (vikhyat187) [reply test/integration/users.test.js:589]: Fetch the user using the userId.

Comment (endeavourmonk) [reply test/integration/users.test.js:589]: We are doing this already.

File: test/integration/users.test.js (hunk 12 above)
Comment (heyrandhir) [reply test/integration/users.test.js:591]: Yes I think its a good idea to test for the fields present. it makes the tests strong. regarding `what fields need to be present` you need to check what fields you have added for the user you are adding in the test.

Comment (endeavourmonk) [reply test/integration/users.test.js:599]: added fields for user in test.

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: test/integration/users.test.js [hunk 13]
```diff
@@ -585,6 +574,48 @@ describe("Users", function () {
...
+          expect(res.body.message).to.equal("User returned successfully!");
+          expect(res.body.user).to.be.a("object");
+          expect(res.body.user).to.have.property("username");
+          expect(res.body.user).to.have.property("first_name");
+          expect(res.body.user).to.have.property("last_name");
+          expect(res.body.user).to.have.property("yoe");
+          expect(res.body.user).to.have.property("linkedin_id");
+          expect(res.body.user).to.have.property("github_id");
+          expect(res.body.user).to.have.property("isMember");
+          expect(res.body.user).to.have.property("roles");
```
Comment (heyrandhir) [inline test/integration/users.test.js:597]: can you try to achieve this in a single expect statement ?

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: test/integration/users.test.js (hunk 13 above)
Comment (endeavourmonk) [reply test/integration/users.test.js]: done.

File: controllers/users.js [hunk 14]
```diff
@@ -70,6 +70,36 @@ const getUsers = async (req, res) => {
...
+      } catch (error) {
+        logger.error(`Error while fetching user: ${error}`);
+        return res.boom.serverUnavailable(SOMETHING_WENT_WRONG);
+      }
+
+      if (!result.userExists) {
+        return res.boom.notFound("User doesn't exist");
+      }
+
+      const { ...user } = result.user;
```
Comment (heyrandhir) [inline controllers/users.js:88]:
```suggestion
      const user = { ...result.user };
```

**Review by heyrandhir — CHANGES_REQUESTED** [review]

**Review by heyrandhir — APPROVED** [review]

File: test/integration/users.test.js (hunk 10 above)
Comment (vikhyat187) [reply test/integration/users.test.js:589]: @RitikJaiswal75 has asked to fetch the data from db, we are only getting hardcoded values.

File: test/integration/users.test.js [hunk 15]
```diff
@@ -585,6 +574,48 @@ describe("Users", function () {
...
+            "isMember",
+            "roles",
+          ]);
+          return done();
+        });
+    });
+    it("Should return 404 if user not Found", function (done) {
+      chai
+        .request(app)
+        .get(`/users/?id=ritiksuserId`)
```
Comment (vikhyat187) [inline test/integration/users.test.js:606]: NIT: use a dummy userId, not the Ritik's Id.

File: test/integration/users.test.js (hunk 10 above)
Comment (RitikJaiswal75) [reply test/integration/users.test.js:589]: How do i check that the data returned has the same user id as the user id passed in query params

Comment (vikhyat187) [reply test/integration/users.test.js:589]: We are not querying the db here, just the API is being called.

`await userQuery.fetchUser({ userId: id });` this is how we query database

**Review by vikhyat187 — CHANGES_REQUESTED** [review]

File: test/integration/users.test.js (hunk 15 above)
Comment (endeavourmonk) [reply test/integration/users.test.js]: updated

File: test/integration/users.test.js (hunk 10 above)
Comment (endeavourmonk) [reply test/integration/users.test.js:596]: modified test as required.

File: controllers/users.js [hunk 16]
```diff
@@ -70,6 +70,36 @@ const getUsers = async (req, res) => {
...
+      }
+
+      const user = { ...result.user };
+      try {
+        delete user.phone;
+        delete user.email;
+      } catch (error) {
+        logger.error(`Error while formatting phone and email: ${error}`);
+        return res.boom.badImplementation("Error while formatting phone and email");
+      }
```
Comment (heyrandhir) [inline controllers/users.js:95]: If the property doesn't exist, the delete operation simply does nothing and doesn't throw any error. we dont need this try catch block

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: controllers/users.js (hunk 16 above)
Comment (endeavourmonk) [reply controllers/users.js]: removed try catch block

File: test/integration/users.test.js [hunk 17]
```diff
@@ -585,6 +575,54 @@ describe("Users", function () {
...
+        .get(`/users/?id=${testUser.id}`)
+        .end((err, res) => {
+          if (err) {
+            return done(err);
+          }
+          expect(res).to.have.status(200);
+          expect(res.body).to.be.a("object");
+          expect(res.body.message).to.equal("User returned successfully!");
+          expect(res.body.user).to.be.a("object");
+          expect(Object.keys(res.body.user)).to.include.members([
```
[the remark this answers is not in the document]
Comment (endeavourmonk) [reply test/integration/users.test.js:602]: done now veryfying the id of the response to be same as testUser's id. And also clearing db after each test

**Review by heyrandhir — APPROVED** [review]

**Review by vikhyat187 — DISMISSED** [review]

File: test/integration/users.test.js [hunk 18]
```diff
@@ -585,6 +575,60 @@ describe("Users", function () {
     });
   });
 
+  describe("GET /users/?id", function () {
+    let testUser;
+    beforeEach(async function () {
+      await addOrUpdate(userData[0], userId);
+      const { user } = await users.fetchUser({ userId });
```
Comment (RitikJaiswal75) [inline test/integration/users.test.js:582]: why are we fetching the user in before each? isn't this specific to a single test or is it required everywhere in this describe?

File: test/integration/users.test.js [hunk 19]
```diff
@@ -585,6 +575,49 @@ describe("Users", function () {
...
   });
 
+  describe("GET /users/?id", function () {
+    afterEach(async function () {
+      await cleanDb();
+    });
+
+    it("Should return given user by id", async function () {
+      await addOrUpdate(userData[0], userId);
+      const { user: testUser } = await users.fetchUser({ userId });
```
Comment (RitikJaiswal75) [inline test/integration/users.test.js:585]: is this assigning data from db or directly assigning the test user to user?

Comment (endeavourmonk) [reply test/integration/users.test.js:585]: user property of the object returned by the fetchUser method to a variable called testUser.

**Comment by vikhyat187** [comment]
@endeavourmonk please update the API contracts and can you also follow up with @sahsisunny he already made changes for website-dashboard https://github.com/Real-Dev-Squad/website-dashboard/pull/342, once get this backend on staging verify it and get it on prod, we can get the dashboard site change deployed.

**Review by vikhyat187 — APPROVED** [review]

**Review by heyrandhir — APPROVED** [review]

**Review by RitikJaiswal75 — DISMISSED** [review]
LGTM

File: controllers/users.js [hunk 20]
```diff
@@ -70,6 +70,31 @@ const getUsers = async (req, res) => {
...
+        return res.boom.serverUnavailable(SOMETHING_WENT_WRONG);
+      }
+
+      if (!result.userExists) {
+        return res.boom.notFound("User doesn't exist");
+      }
+
+      const user = { ...result.user };
+      delete user.phone;
+      delete user.email;
```
Comment (prakashchoudhary07) [inline controllers/users.js:90]: I think we should remove token and chain code also
REF https://github.com/Real-Dev-Squad/website-backend/blob/a4952edc866ccbb6d9a8645481e9304e67adc4ae/models/users.js#L167-L174

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: controllers/users.js (hunk 20 above)
Comment (endeavourmonk) [reply controllers/users.js:90]: okay

File: models/users.js [hunk 21]
```diff
@@ -170,8 +170,6 @@ const fetchPaginatedUsers = async (query) => {
         ...doc.data(),
         phone: undefined,
         email: undefined,
-        tokens: undefined,
-        chaincode: undefined,
```
Comment (prakashchoudhary07) [inline models/users.js:174]: Why are these been removed here?

File: controllers/users.js (hunk 20 above)
Comment (endeavourmonk) [reply controllers/users.js:94]: > I think we should remove token and chain code also
> REF

You asked so, I removed them

File: models/users.js (hunk 21 above)
Comment (endeavourmonk) [reply models/users.js:174]: In the previous comment you told "I think we should remove token and chain code also
REF" so I removed them

File: test/integration/users.test.js [hunk 22]
```diff
@@ -592,6 +582,50 @@ describe("Users", function () {
...
```
Comment (prakashchoudhary07) [inline test/integration/users.test.js:592]: NIT
```suggestion
      const { userId } = await addOrUpdate(userData[0]);
```

**Review by prakashchoudhary07 — APPROVED** [review]

**Pull request merged**